src/app/learnocaml_index_main.ml

```diff
@@ -85,8 +87,76 @@ let (exercise_filter_signal: string option React.signal), set_exercise_filter =
 let (exercise_sort_signal: exercise_ordering React.signal), set_exercise_sort =
   React.S.create By_category
 
+let (expand_state: string list React.signal), set_expand_state =
+  React.S.create []
+
+let encode str =
+  Re.Pcre.substitute ~rex:(Re.Pcre.regexp ",") ~subst:(fun _ -> "-c") (
+      Re.Pcre.substitute ~rex:(Re.Pcre.regexp "&") ~subst:(fun _ -> "-a") (
+          Re.Pcre.substitute ~rex:(Re.Pcre.regexp "=") ~subst:(fun _ -> "-e") (
+              Re.Pcre.substitute ~rex:(Re.Pcre.regexp "-") ~subst:(fun _ -> "--") str)))
+let decode str =
+  Re.Pcre.substitute ~rex:(Re.Pcre.regexp "--") ~subst:(fun _ -> "-") (
+      Re.Pcre.substitute ~rex:(Re.Pcre.regexp "-e") ~subst:(fun _ -> "=") (
+          Re.Pcre.substitute ~rex:(Re.Pcre.regexp "-a") ~subst:(fun _ -> "&") (
+              Re.Pcre.substitute ~rex:(Re.Pcre.regexp "-c") ~subst:(fun _ -> ",") str)))
+
+let rec update_expand ?value fragment =
+  match value with
+  | Some v ->
+     if (List.mem_assoc "expand" fragment) then
+       match fragment with
+       | [] ->
+          update_expand ~value:v []
+       | ("expand", x)::t ->
+          let expand_list = Re.Pcre.split ~rex:(Re.Pcre.regexp ",") x in
+          let new_expand_list =
+            if (List.mem v expand_list) then
+              begin
+                let filtered_list = List.filter (fun x -> x <> v) expand_list in
+                if (filtered_list = []) then [""] else filtered_list
+              end
+            else
+              begin
+                if (expand_list = [""]) then [v] else v::expand_list
+              end
+          in
+          let join l =
+            match l with
+            | [] -> ""
+            | h::t -> List.fold_left (fun acc x -> acc ^ "," ^ x) h t
+          in
+          let joined_expand = join new_expand_list in
+          set_expand_state new_expand_list;
+          (*if (joined_expand <> "") then*) ("expand", joined_expand)::t (*else t*)
+       | h::t ->
+          h::(update_expand ~value:v t)
+     else
+       begin
+         set_expand_state [v];
+         fragment@[("expand",v)]
+       end
+  | None ->
+     set_expand_state [];
+     List.filter (fun (k, _) -> k <> "expand") fragment
+
+let update_sort value fragment =
+  let filtered_fragment = List.remove_assoc "sort" (update_expand fragment) in
+  filtered_fragment@[("sort",value)]
+
+let update_fragment key value =
+  let fragment = Js_utils.parse_fragment () in
+  let filtered_fragment =
+    if (key = "expand") then
+      let v = (Uri.pct_encode (encode value)) in
```

Comment:
This part of the code calls the `encode` function. Where is called the `decode` function?